src/software/backend/output/wifi/communication/protobuf_primitive_visitor.cpp

```diff
@@ -0,0 +1,151 @@
+    prim_msg->set_prim_type(FirmwarePrimitiveType::DIRECT_WHEELS);
+    prim_msg->set_parameter1(
+        static_cast<double>(direct_wheels_primitive.getWheel0Power()));
+    prim_msg->set_parameter2(
+        static_cast<double>(direct_wheels_primitive.getWheel1Power()));
+    prim_msg->set_parameter3(
+        static_cast<double>(direct_wheels_primitive.getWheel2Power()));
+    prim_msg->set_parameter4(
+        static_cast<double>(direct_wheels_primitive.getWheel3Power()));
+    prim_msg->set_extra_bits(direct_wheels_primitive.getDribblerRPM() *
+                             DRIBBLER_RPM_TO_RADIO_CONVERSION_FACTOR);
+}
+
+void ProtobufPrimitiveVisitor::visit(const DribblePrimitive &dribble_primitive)
+{
+    prim_msg = PrimitiveMsg();
+    prim_msg->set_prim_type(FirmwarePrimitiveType::DRIBBLE);
+    prim_msg->set_parameter1(dribble_primitive.getDestination().x() *
+                             MILLIMETERS_PER_METER);
+    prim_msg->set_parameter2(dribble_primitive.getDestination().y() *
+                             MILLIMETERS_PER_METER);
+    prim_msg->set_parameter3(dribble_primitive.getFinalAngle().toRadians() *
+                             CENTIRADIANS_PER_RADIAN);
+    // For this primitive, we don't divide the RPM
+    prim_msg->set_parameter4(dribble_primitive.getRpm());
+    prim_msg->set_extra_bits(dribble_primitive.isSmallKickAllowed());
+}
+
+void ProtobufPrimitiveVisitor::visit(const KickPrimitive &kick_primitive)
+{
+    prim_msg = PrimitiveMsg();
+    prim_msg->set_prim_type(FirmwarePrimitiveType::SHOOT);
+    prim_msg->set_parameter1(kick_primitive.getKickOrigin().x() * MILLIMETERS_PER_METER);
+    prim_msg->set_parameter2(kick_primitive.getKickOrigin().y() * MILLIMETERS_PER_METER);
+    prim_msg->set_parameter3(kick_primitive.getKickDirection().toRadians() *
+                             CENTIRADIANS_PER_RADIAN);
+    prim_msg->set_parameter4(kick_primitive.getKickSpeed() * MILLIMETERS_PER_METER);
+    prim_msg->set_extra_bits(static_cast<uint32_t>(2 | 0));
+}
+
+void ProtobufPrimitiveVisitor::visit(const MovePrimitive &move_primitive)
+{
+    prim_msg = PrimitiveMsg();
+    prim_msg->set_prim_type(FirmwarePrimitiveType::MOVE);
+    prim_msg->set_parameter1(move_primitive.getDestination().x() * MILLIMETERS_PER_METER);
+    prim_msg->set_parameter2(move_primitive.getDestination().y() * MILLIMETERS_PER_METER);
+    prim_msg->set_parameter3(move_primitive.getFinalAngle().toRadians() *
+                             CENTIRADIANS_PER_RADIAN);
+    prim_msg->set_parameter4(move_primitive.getFinalSpeed() * MILLIMETERS_PER_METER);
+    prim_msg->set_slow(move_primitive.getMoveType() == MoveType::SLOW);
+
+    uint32_t extra_bits = 0;
+    extra_bits |= (move_primitive.getAutoKickType() == AUTOKICK) * 0x01;
+    extra_bits |= (move_primitive.getDribblerEnable() == DribblerEnable::ON) * 0x02;
+    extra_bits |= (move_primitive.getAutoKickType() == AUTOCHIP) * 0x04;
+    prim_msg->set_extra_bits(extra_bits);
```

Comment:
bit messy, would like to see some abstraction around creating prim_msg so the visit function is much cleaner.

Reply:
This logic is going to have to live somewhere, even with a function like that, you have to add checks to do specific things based on the msg type internally. 

It will just change where the code is.